File: app/src/main/java/com/android/serialport2/CrashHandler.java
package com.android.serialport2;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Environment;
import android.os.Looper;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.reflect.Field;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;


public class CrashHandler implements Thread.UncaughtExceptionHandler {
    private Thread.UncaughtExceptionHandler mDefaultHandler;
    @SuppressLint("StaticFieldLeak")
    private static final CrashHandler INSTANCE = new CrashHandler();
    private Context mContext;
    private final Map<String, String> appInformation = new HashMap<>();

    @SuppressLint("SimpleDateFormat")
    private final DateFormat formatter = new SimpleDateFormat("yyyyMMdd_HHmmss");
    private File logDir;

    private CrashHandler() {
    }

    public static CrashHandler getInstance() {
        return INSTANCE;
    }

    public void init(Context context) {
        mContext = context;
        logDir = mContext.getExternalFilesDir("LOG");
        mDefaultHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(this);
    }

    @Override
    public void uncaughtException(Thread thread, Throwable ex) {
        if (!handleException(ex) && mDefaultHandler != null) {
            mDefaultHandler.uncaughtException(thread, ex);
        } else {
            try {
                Thread.sleep(3000);
            } catch (InterruptedException ignored) {
            }
            android.os.Process.killProcess(android.os.Process.myPid());
            System.exit(1);
        }
    }

    private boolean handleException(Throwable ex) {
        if (ex == null) {
            return false;
        }
        new Thread(() -> {
            Looper.prepare();
            Toast.makeText(mContext, "很抱歉," + mContext.getPackageName() + "程序出现异常,正在收集日志，即将退出", Toast.LENGTH_SHORT).show();
            Looper.loop();
        }).start();
        collectDeviceInfo(mContext);
        saveCrashInfo2File(ex);
        return true;
    }

    public void collectDeviceInfo(Context ctx) {
        try {
            PackageManager pm = ctx.getPackageManager();
            PackageInfo pi = pm.getPackageInfo(ctx.getPackageName(), PackageManager.GET_ACTIVITIES);
            if (pi != null) {
                String versionName = pi.versionName == null ? "null" : pi.versionName;
                String versionCode = pi.versionCode + "";
                appInformation.put("versionName", versionName);
                appInformation.put("versionCode", versionCode);
            }
        } catch (PackageManager.NameNotFoundException ignored) {
        }
        Field[] fields = Build.class.getDeclaredFields();
        for (Field field : fields) {
            try {
                field.setAccessible(true);
                appInformation.put(field.getName(), Objects.requireNonNull(field.get(null)).toString());
            } catch (Exception ignored) {
            }
        }
    }

    private void saveCrashInfo2File(Throwable ex) {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, String> entry : appInformation.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            sb.append(key).append("=").append(value).append("\n");
        }

        Writer writer = new StringWriter();
        PrintWriter printWriter = new PrintWriter(writer);
        ex.printStackTrace(printWriter);
        Throwable cause = ex.getCause();
        while (cause != null) {
            cause.printStackTrace(printWriter);
            cause = cause.getCause();
        }
        printWriter.close();
        String result = writer.toString();
        sb.append(result);
        try {
            long timestamp = System.currentTimeMillis();
            String time = formatter.format(new Date());
            String fileName = time + "_" + timestamp + ".txt";
            if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
                File file = new File(logDir, fileName);
                FileOutputStream fos = new FileOutputStream(file);
                fos.write(sb.toString().getBytes());
                fos.close();
            }
        } catch (Exception ignored) {
        }
    }
}  

File: app/src/main/java/com/android/serialport2/other/CrashHandler.kt
package com.android.serialport2.other

import android.content.Context
import android.os.Build
import android.os.Looper
import android.os.Process
import android.widget.Toast
import com.android.serialport2.BuildConfig
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.io.Writer
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.system.exitProcess

class CrashHandler : Thread.UncaughtExceptionHandler {
    companion object {
        val instance: CrashHandler = CrashHandler()
    }

    private var mDefaultHandler: Thread.UncaughtExceptionHandler? = null
    private var mContext: Context? = null
    private val infos: MutableMap<String, String> = mutableMapOf()
    private var logDir: File? = null

    private val lock = Any()

    @Volatile
    private var installed: Boolean = false

    fun init(context: Context) {
        val appContext = context.applicationContext
        synchronized(lock) {
            if (installed) return
            installed = true
            mContext = appContext
            logDir = resolveLogDir(appContext)
            mDefaultHandler = Thread.getDefaultUncaughtExceptionHandler()
            Thread.setDefaultUncaughtExceptionHandler(this)
        }
    }

    override fun uncaughtException(thread: Thread, ex: Throwable) {
        try {
            if (!handleException(ex) && mDefaultHandler != null) {
                mDefaultHandler?.uncaughtException(thread, ex)
                return
            } else {
                try {
                    Thread.sleep(3000)
                } catch (_: InterruptedException) {
                }
                Process.killProcess(Process.myPid())
                exitProcess(1)
            }
        } catch (_: Throwable) {
        }
    }

    private fun handleException(ex: Throwable?): Boolean {
        if (ex == null) return false
        val ctx = mContext ?: return false
        try {
            Thread {
                try {
                    Looper.prepare()
                    Toast.makeText(
                        ctx,
                        "串口工具异常退出,正在收集日志",
                        Toast.LENGTH_LONG
                    )
                        .show()
                    Looper.loop()
                } catch (_: InterruptedException) {
                }
            }.start()
        } catch (_: Throwable) {
        }
        collectDeviceInfo()
        saveCrashInfo2File(ex)
        return true
    }

    fun collectDeviceInfo() {
        synchronized(lock) {
            infos.clear()
            infos["versionName"] = BuildConfig.VERSION_NAME
            infos["versionCode"] = "${BuildConfig.VERSION_CODE}"
            // 设备信息：使用白名单，避免全反射导致日志过大/含敏感字段
            infos["sdkInt"] = Build.VERSION.SDK_INT.toString()
            infos["release"] = Build.VERSION.RELEASE ?: "unknown"
            infos["incremental"] = Build.VERSION.INCREMENTAL ?: "unknown"
            infos["brand"] = Build.BRAND ?: "unknown"
            infos["manufacturer"] = Build.MANUFACTURER ?: "unknown"
            infos["model"] = Build.MODEL ?: "unknown"
            infos["device"] = Build.DEVICE ?: "unknown"
            infos["product"] = Build.PRODUCT ?: "unknown"
            infos["hardware"] = Build.HARDWARE ?: "unknown"
            infos["fingerprint"] = Build.FINGERPRINT ?: "unknown"
        }
    }

    private fun saveCrashInfo2File(ex: Throwable) {
        val sb = StringBuilder()
        sb.append("time=").append(nowTimestamp()).append('\n')
        sb.append("thread=").append(Thread.currentThread().name).append('\n')
        val infoSnapshot: Map<String, String> = synchronized(lock) { HashMap(infos) }
        for ((key, value) in infoSnapshot.entries) {
            sb.append(key).append('=').append(value).append('\n')
        }
        val writer: Writer = StringWriter()
        val printWriter = PrintWriter(writer)
        ex.printStackTrace(printWriter)
        var cause = ex.cause
        while (cause != null) {
            cause.printStackTrace(printWriter)
            cause = cause.cause
        }
        printWriter.close()
        val result = writer.toString()
        sb.append(result)
        val time = nowTimestamp()
        val fileName = "$time.log"
        val primaryFile = File(logDir, fileName)
        tryWriteText(primaryFile, sb.toString())
    }

    private fun resolveLogDir(ctx: Context): File {
        val ext = ctx.getExternalFilesDir("LOG")
        val dir = ext ?: File(ctx.filesDir, "LOG")
        if (!dir.exists()) {
            try {
                dir.mkdirs()
            } catch (_: Exception) {
            }
        }
        return dir
    }

    private fun nowTimestamp(): String {
        return try {
            SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(Date())
        } catch (_: Throwable) {
            Date().time.toString()
        }
    }

    private fun tryWriteText(file: File, content: String) {
        return try {
            file.parentFile?.let {
                if (!it.exists()) {
                    try {
                        it.mkdirs()
                    } catch (_: Exception) {
                    }
                }
            }
            file.outputStream().use { it.write(content.toByteArray()) }
        } catch (_: Throwable) {
        }
    }
}